vault/resource_ad_secret_backend.go

```diff
@@ -0,0 +1,645 @@
+		},
+		"ttl": {
+			Type:        schema.TypeInt,
+			Optional:    true,
+			Computed:    true,
+			Description: `In seconds, the default password time-to-live.`,
+		},
+		"upndomain": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: `Enables userPrincipalDomain login with [username]@UPNDomain (optional)`,
+		},
+		"url": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: `LDAP URL to connect to (default: ldap://127.0.0.1). Multiple URLs can be specified by concatenating them with commas; they will be tried in-order.`,
+		},
+		"use_pre111_group_cn_behavior": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Computed:    true,
+			Description: `In Vault 1.1.1 a fix for handling group CN values of different cases unfortunately introduced a regression that could cause previously defined groups to not be found due to a change in the resulting name. If set true, the pre-1.1.1 behavior for matching group CNs will be used. This is only needed in some upgrade scenarios for backwards compatibility. It is enabled by default if the config is upgraded but disabled by default on new configurations.`,
+		},
+		"use_token_groups": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Description: `If true, use the Active Directory tokenGroups constructed attribute of the user to find the group memberships. This will find all security groups including nested ones.`,
+		},
+		"userattr": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: `Attribute used for users (default: cn)`,
+		},
+		"userdn": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: `LDAP domain to use for users (eg: ou=People,dc=example,dc=org)`,
+		},
+	}
+	return &schema.Resource{
+		Create: createConfigResource,
+		Update: updateConfigResource,
+		Read:   readConfigResource,
+		Delete: deleteConfigResource,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: fields,
+	}
+}
+
+func createConfigResource(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	backend := d.Get("backend").(string)
+	description := d.Get("description").(string)
+	defaultTTL := d.Get("default_lease_ttl_seconds").(int)
+	local := d.Get("local").(bool)
+	maxTTL := d.Get("max_lease_ttl_seconds").(int)
+
+	log.Printf("[DEBUG] Mounting AD backend at %q", backend)
+	err := client.Sys().Mount(backend, &api.MountInput{
+		Type:        "ad",
+		Description: description,
+		Local:       local,
+		Config: api.MountConfigInput{
+			DefaultLeaseTTL: fmt.Sprintf("%ds", defaultTTL),
+			MaxLeaseTTL:     fmt.Sprintf("%ds", maxTTL),
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("error mounting to %q: %s", backend, err)
+	}
+
+	log.Printf("[DEBUG] Mounted AD backend at %q", backend)
+	d.SetId(backend)
+
+	data := map[string]interface{}{}
+	if v, ok := d.GetOkExists("anonymous_group_search"); ok {
```

Comment:
Any reason not to use a loop over a slice of strings here?

Definitely not for this PR, but a wider point:
Also I know it's consistent with the style elsewhere, but it scares me that we duplicate all these strings between the schema and implementation. It seems like an easy win to just have some consts, although given everything is currently in the `vault` package, perhaps we would want to split into multiple packages to prevent massive top-level name noise.

Reply:
I think this would make a good cleanup across the board. Something we could target in the future.